### PR TITLE
fix RHP animation on Safari

### PR DESCRIPTION
**Pull request by Kicu**
### Details
 - hotfix for the bug with RHP screens
 - Stop using interpolator from `react-navigation` for RHP because it seems to be the source of the bug on Safari
 - instead use an interpolator that was already written and used in another place

### Fixed Issues

$ https://github.com/Expensify/App/issues/44765
PROPOSAL:

### Tests

- verify opening RHP and navigating within it
- flow: FAB -> Assign Task -> Next
- flow: Report -> ReportDetails -> ShareScreen 
- flow: Search -> click on any row -> ReportDetails (click on header inside RHP)

### QA Steps

- verify opening RHP and navigating within it
- flow: FAB -> Assign Task -> Next
- flow: Report -> ReportDetails -> ShareScreen 
- flow: Search -> click on any row -> ReportDetails (click on header inside RHP)

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

#### Chrome
https://github.com/Expensify/App/assets/3929868/4695b37f-010b-40ff-91f1-532759f3c1d7

#### Safari
https://github.com/Expensify/App/assets/3929868/da379567-0cba-4db5-8e22-c9b90b787f7a

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by Kicu** [comment]
@mountiny this is the fix, please take a look - perhaps we could run a build here and test it? 
I have tested locally on Safari and did not get the bug

CC @ShridharGoel @s77rt

**Comment by s77rt** [comment]
cc @ishpaul777 (https://github.com/Expensify/App/issues/44765#issuecomment-2209428056)

**Comment by Kicu** [comment]
I see, it's a shame that this doesn't solve the bug.

However the funny thing is that this solves it for me 100%. I literally cannot trigger the bug anymore which means I can't try to fix it anymore 😓 

Check my video. Interestingly if you specifically look at the period around 0:25 - 0:40 you can see the very similar faulty animation - but for me it always comes back to correct layout. However indeed there is something there still problematic. But I still think its these interpolators that break it.

https://github.com/Expensify/App/assets/3929868/a68fd607-2b2f-453d-8244-8c8898efba34

Not sure what to do next, since I cannot reproduce it anymore - either via clicking or via keyboard and I tried multiple types of RHP modals.
**One more thing to try:** There is a few more places where we use `CardStyleInterpolators.forHorizontalIOS` - people that still see the bug could try to remove it everywhere and see if this helps?

Btw I'm on mac M3 Pro, Sonoma v14.4.1, Safari v17.14.1

**Comment by mountiny** [comment]
I can still reliably repro on this branch as well when going from the task title to confirmation page
https://github.com/Expensify/App/assets/36083550/95bdff9d-5c11-49bb-b712-fa2cde2e31de


@Kicu nobody in SWM can reproduce this?

**Comment by Kicu** [comment]
This PR is now using the custom interpolators only for Safari as a temporary solution, until we can fix all the animations correctly.

File: src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.tsx [hunk 1]
```diff
@@ -18,10 +21,17 @@ const Stack = createStackNavigator<RightModalNavigatorParamList>();
...
     const styles = useThemeStyles();
+    const styleUtils = useStyleUtils();
     const {isSmallScreenWidth} = useWindowDimensions();
     const screenOptions = useMemo(() => ModalNavigatorScreenOptions(styles), [styles]);
     const isExecutingRef = useRef<boolean>(false);
 
+    // The .forHorizontalIOS interpolator from `@react-navigation` is misbehaving on Safari, so we override it with Expensify custom interpolator
+    if (isSafari()) {
+        const customInterpolator = createModalCardStyleInterpolator(styleUtils);
+        screenOptions.cardStyleInterpolator = (props: StackCardInterpolationProps) => customInterpolator(isSmallScreenWidth, false, false, props);
```
[the remark this answers is not in the document]
Comment (Kicu) [reply src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.tsx:32]: sure 👍  will do it in ~10-15minutes

**Comment by Kicu** [comment]
@ishpaul777 PR updated, pushed a with commit with usememo

File: src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts [hunk 2]
```diff
@@ -1,19 +1,32 @@
-import type {StackNavigationOptions} from '@react-navigation/stack';
+import type {StackCardInterpolationProps, StackNavigationOptions} from '@react-navigation/stack';
 import {CardStyleInterpolators} from '@react-navigation/stack';
 import {useMemo} from 'react';
+import useStyleUtils from '@hooks/useStyleUtils';
 import useThemeStyles from '@hooks/useThemeStyles';
+import useWindowDimensions from '@hooks/useWindowDimensions';
+import {isSafari} from '@libs/Browser';
```
[the remark this answers is not in the document]
Comment (Kicu) [reply src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts:7]: sorry forgot to change this, lets just keep as is - this code is supposed to go away soon when we have a better solution. Will keep this in mind for future

**Review by ishpaul777 — APPROVED** [review]
we just need to apply suggestion rest looks good! Approving since I have go offline for few hours. @mountiny Can you please handle it forward.

File: src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.tsx [hunk 3]
```diff
@@ -18,9 +21,19 @@ const Stack = createStackNavigator<RightModalNavigatorParamList>();
...
+    const styleUtils = useStyleUtils();
     const {isSmallScreenWidth} = useWindowDimensions();
-    const screenOptions = useMemo(() => ModalNavigatorScreenOptions(styles), [styles]);
     const isExecutingRef = useRef<boolean>(false);
+    const screenOptions = useMemo(() => {
+        const options = ModalNavigatorScreenOptions(styles);
+        // The .forHorizontalIOS interpolator from `@react-navigation` is misbehaving on Safari, so we override it with Expensify custom interpolator
+        if (isSafari()) {
+            const customInterpolator = createModalCardStyleInterpolator(styleUtils);
+            screenOptions.cardStyleInterpolator = (props: StackCardInterpolationProps) => customInterpolator(isSmallScreenWidth, false, false, props);
```
[the remark this answers is not in the document]
Comment (Kicu) [reply src/libs/Navigation/AppNavigator/Navigators/RightModalNavigator.tsx]: done, thank you for catching this, I was kinda rushing it 😓

**Review by ishpaul777 — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
Thanks!

**Pull request merged**